src/hotspot/share/opto/loopTransform.cpp

```diff
@@ -2522,165 +2522,222 @@ void PhaseIdealLoop::add_constraint(jlong stride_con, jlong scale_con, Node* off
   }
 }
 
+//----------------------------------is_iv------------------------------------
+// Return true if exp is the value (of type bt) of the given induction var.
+// This grammar of cases is recognized, where X is I|L according to bt:
+//    VIV[iv] = iv | (CastXX VIV[iv]) | (ConvI2X VIV[iv])
 bool PhaseIdealLoop::is_iv(Node* exp, Node* iv, BasicType bt) {
-  if (exp == iv) {
+  exp = exp->uncast();
+  if (exp == iv && iv->bottom_type()->isa_integer(bt)) {
     return true;
   }
 
-  if (bt == T_LONG && iv->bottom_type()->isa_int() && exp->Opcode() == Op_ConvI2L && exp->in(1) == iv) {
+  if (bt == T_LONG && iv->bottom_type()->isa_int() && exp->Opcode() == Op_ConvI2L && exp->in(1)->uncast() == iv) {
     return true;
   }
   return false;
 }
 
 //------------------------------is_scaled_iv---------------------------------
-// Return true if exp is a constant times an induction var
-bool PhaseIdealLoop::is_scaled_iv(Node* exp, Node* iv, jlong* p_scale, BasicType bt) {
-  exp = exp->uncast();
-  assert(bt == T_INT || bt == T_LONG, "unexpected int type");
-  if (is_iv(exp, iv, bt)) {
+// Return true if exp is a constant times the given induction var (of type bt).
+// The multiplication must be done in full precision (exactly of type bt).
```

Comment:
Now that the short-scale detection is factored up here, the comment should change also.
I've also tweaked the pseudocode grammar to be (I think) more correct.
```
 // The multiplication is either done in full precision (exactly of type bt),
 // or else bt is T_LONG but iv is scaled using 32-bit arithmetic followed by a ConvI2L.
 // This grammar of cases is recognized, where X is I|L according to bt:
 //    SIV[iv] = VIV[iv] | (CastXX SIV[iv])
 //            | (MulX VIV[iv] ConX) | (MulX ConX VIV[iv])
 //            | (LShiftX VIV[iv] ConI)
 //            | (ConvI2L SIV[iv])  -- a "short-scale" can occur here; note recursion
 //            | (SubX 0 SIV[iv])  -- same as MulX(iv, -scale); note recursion
 // On success, the constant scale value is stored back to ret_scale.
 // The value (*p_short_scale) reports if such a ConvI2L conversion was present.
```